### DISCUSSION TITLE
Body's world center changes after calling setMass(MassType type)

### DISCUSSION THREAD

**Discussion by Pecolyte** [dyn4j/dyn4j#281] [General]
Not sure if this is intentional. But calling setMass(MassType type) after modifying the fixtures, as the document says, will change the world center of the body, which seems to be the caculated center point of all fixtures, even if the body hasn't been added to the world. On the other hand, calling setMass(Mass mass) on the body would have no such effect. I'm guessing that caculating the mass of the body causes this?

Fixtures added after calling the setMass(MassType type) method won't affect. On the other word, if I call the setMass(MassType type) on the very first after creating the body, and then add the fixtures, the world center will remain on (0, 0). But this is not recommended in the document.

So I found this a little confusing.  The world center behaves differently between setMass(MassType type) and setMass(Mass mass), causing trouble for drawing on the canvas. Did I misunderstand something and using it the wrong way?

BTW I've been using the version 5.0.1.

**Comment by wnbittle** [comment]
> Not sure if this is intentional. But calling setMass(MassType type) after modifying the fixtures, as the document says, will change the world center of the body, which seems to be the caculated center point of all fixtures, even if the body hasn't been added to the world. On the other hand, calling setMass(Mass mass) on the body would have no such effect. I'm guessing that caculating the mass of the body causes this?

This is intentional.  The mass, inertia, and center of mass of a `Body` are independent of what `World` it will be added to and you _want_ those values set before it's simulated.  The `setMass(MassType)` is used to _calculate_ the mass, inertia, and center of mass of a `Body` based on it's `Fixture`s (and `Shape`), density, and their location within the `Body`.  The `setMass(Mass)` method is used to _set_ the mass, inertia, and center of mass manually.  If you are using the meter-kilogram-seconds system (which you should), then calling `setMass(MassType)` is the best way to ensure that a `Body` reacts in a reasonable way to physical events.

> Fixtures added after calling the setMass(MassType type) method won't affect. On the other word, if I call the setMass(MassType type) on the very first after creating the body, and then add the fixtures, the world center will remain on (0, 0). But this is not recommended in the document.

This is intentional as well.  The `setMass(MassType)` method only knows what you've added to the `Body`.  If nothing has been added, then the mass is 0, the inertia is 0, and the center of mass is (0,0).  If you don't like the `Mass` properties calculated by the `setMass(MassType)` method, you can just override it when you are done.  For example, here I want to keep the mass/inertia numbers, but I want to set the center to (0,0).
```java
// use the calculation
body.setMass(MassType.NORMAL);
// override the center of mass
Mass originalMass = body.getMass();
Mass newMass = new Mass(originalMass.getMass(), originalMass.getInertia(), new Vector2(0, 0));
// set the mass manually
body.setMass(newMass);
```
This isn't recommended because (0,0) _isn't_ the physically correct center mass. Changing the center of mass like this may cause the simulation to produce unexpected behavior.

> So I found this a little confusing. The world center behaves differently between setMass(MassType type) and setMass(Mass mass), causing trouble for drawing on the canvas. Did I misunderstand something and using it the wrong way?

If you need some help drawing, [take a look at the render code in the samples project](https://github.com/dyn4j/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/framework/Graphics2DRenderer.java).

**Comment by Pecolyte** [comment]
It makes more sense now, thank you for the reply!

What I've done is drawing the fixtures using the world center and the fixtures' transformed local centers, which was performed right after added to the body. With that, the world center is still at (0, 0), and then the mess was set but the drawing thread is still considering the world center being at (0, 0), and the local center is recaculated to match the new world center, which caused the problem.

With this being clear, I'm updating both of the centers every frame, which works fine. Thanks again!
